wallet/wallet-controller/src/sync/mod.rs

```diff
@@ -0,0 +1,172 @@
+
+    let node_best_block =
+        rpc_client.get_best_block_id().await.map_err(SyncError::UnexpectedRpcError)?;
+
+    let common_block_opt = rpc_client
+        .get_last_common_block(wallet_block_id, node_best_block)
+        .await
+        .map_err(SyncError::UnexpectedRpcError)?;
+
+    let (common_block_id, common_block_height) = match common_block_opt {
+        Some(common_block) => common_block,
+        None => {
+            return Ok(Some(SyncEvent::Reset {
+                block_id: chain_config.genesis_block_id(),
+                block_height: BlockHeight::zero(),
+            }));
+        }
+    };
+
+    if common_block_height < wallet_block_height {
+        return Ok(Some(SyncEvent::Reset {
+            block_id: common_block_id,
+            block_height: common_block_height,
+        }));
+    }
+
+    let next_block_height = wallet_block_height.next_height();
+    let next_block_id_opt = rpc_client
+        .get_block_id_at_height(next_block_height)
+        .await
+        .map_err(SyncError::UnexpectedRpcError)?;
+    let next_gen_block_id = match next_block_id_opt {
+        Some(id) => id,
+        None => return Ok(None),
+    };
+    let next_block_id = match next_gen_block_id.classify(chain_config) {
+        common::chain::GenBlockId::Genesis(_) => {
+            return Err(SyncError::UnexpectedGenesisBlock(wallet_block_height))
+        }
+        common::chain::GenBlockId::Block(id) => id,
+    };
+
+    let new_block = rpc_client
+        .get_block(next_block_id)
+        .await
+        .map_err(SyncError::UnexpectedRpcError)?;
+    match new_block {
+        Some(block) => Ok(Some(SyncEvent::NewBlock { block })),
+        // This may fail if there was a reorg after this function was started.
+        // It should resolve normally next time.
+        None => Err(SyncError::BlockNotFound(next_block_id)),
+    }
+}
+
+pub async fn run<T: NodeInterface>(
+    sync_tx: mpsc::Sender<SyncEvent>,
+    chain_config: Arc<ChainConfig>,
+    mut rpc_client: T,
+    mut wallet_block_id: Id<GenBlock>,
+    mut wallet_block_height: BlockHeight,
+) {
+    while !sync_tx.is_closed() {
+        let sync_state = fetch_new_block(
+            &chain_config,
+            &mut rpc_client,
+            wallet_block_id,
+            wallet_block_height,
+        )
+        .await;
+
+        match sync_state {
+            Ok(Some(v)) => {
+                match &v {
+                    SyncEvent::Reset {
+                        block_id,
+                        block_height,
+                    } => {
+                        wallet_block_id = *block_id;
+                        wallet_block_height = *block_height;
+                    }
+                    SyncEvent::NewBlock { block } => {
+                        wallet_block_id = block.header().block_id().into();
+                        wallet_block_height = wallet_block_height.next_height();
```

Comment:
Updating the state locally, as discussed, is a bad idea. The problem here is that if anything goes wrong in the processing or sending, this event loop will be perpetually running for ever, sending blocks the wallet cannot recognize, causing more trouble, because nothing can stop it, ever, other than shutting the program down.

Reply:
Right, thanks. Should be fixed now. Block sync is started from the event loop now.